From 673f425124df882e0eb63ccd52443aeb5d62a2fe Mon Sep 17 00:00:00 2001
From: Rob Evers <revers@redhat.com>
Date: Mon, 23 Jun 2014 21:41:57 -0400
Subject: [scsi] lpfc 10.2.8000.0: Fix FW dump using sysfs

Message-id: <1403559726-20488-16-git-send-email-revers@redhat.com>
Patchwork-id: 84423
O-Subject: [RHEL7.1 e-stor PATCH 15/24] scsi: lpfc 10.2.8000.0: Fix FW dump using sysfs
Bugzilla: 1088574
RH-Acked-by: Ewan Milne <emilne@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Dick Kennedy <rkennedy@Emulex.Com>

Submitted by James Smart James.Smart@emulex.com

Fix FW dump using sysfs

Signed-off-by: James Smart james.smart@emulex.com

http://marc.info/?l=linux-scsi&m=139663393621861&w=2
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/lpfc/lpfc.h b/drivers/scsi/lpfc/lpfc.h
index 94a3caf..f713758 100644
--- a/drivers/scsi/lpfc/lpfc.h
+++ b/drivers/scsi/lpfc/lpfc.h
@@ -640,6 +640,7 @@ struct lpfc_hba {
 #define HBA_DEVLOSS_TMO         0x2000 /* HBA in devloss timeout */
 #define HBA_RRQ_ACTIVE  0x4000 /* process the rrq active list */
 #define HBA_FCP_IOQ_FLUSH 0x8000 /* FCP I/O queues being flushed */
+#define HBA_FW_DUMP_OP  0x10000 /* Skips fn reset before FW dump */
  uint32_t fcp_ring_in_use; /* When polling test if intr-hndlr active*/
  struct lpfc_dmabuf slim2p;
 
diff --git a/drivers/scsi/lpfc/lpfc_attr.c b/drivers/scsi/lpfc/lpfc_attr.c
index 9567b1b..98254ae 100644
--- a/drivers/scsi/lpfc/lpfc_attr.c
+++ b/drivers/scsi/lpfc/lpfc_attr.c
@@ -919,10 +919,15 @@ lpfc_sli4_pdev_reg_request(struct lpfc_hba *phba, uint32_t opcode)
   phba->cfg_sriov_nr_virtfn = 0;
  }
 
+ if (opcode == LPFC_FW_DUMP)
+  phba->hba_flag |= HBA_FW_DUMP_OP;
+
  status = lpfc_do_offline(phba, LPFC_EVT_OFFLINE);
 
- if (status != 0)
+ if (status != 0) {
+  phba->hba_flag &= ~HBA_FW_DUMP_OP;
   return status;
+ }
 
  /* wait for the device to be quiesced before firmware reset */
  msleep(100);
diff --git a/drivers/scsi/lpfc/lpfc_sli.c b/drivers/scsi/lpfc/lpfc_sli.c
index 5f1fc9a..c736002 100644
--- a/drivers/scsi/lpfc/lpfc_sli.c
+++ b/drivers/scsi/lpfc/lpfc_sli.c
@@ -4036,12 +4036,13 @@ lpfc_sli4_brdreset(struct lpfc_hba *phba)
 {
  struct lpfc_sli *psli = &phba->sli;
  uint16_t cfg_value;
- int rc;
+ int rc = 0;
 
  /* Reset HBA */
  lpfc_printf_log(phba, KERN_INFO, LOG_SLI,
-   "0295 Reset HBA Data: x%x x%x\n",
-   phba->pport->port_state, psli->sli_flag);
+   "0295 Reset HBA Data: x%x x%x x%x\n",
+   phba->pport->port_state, psli->sli_flag,
+   phba->hba_flag);
 
  /* perform board reset */
  phba->fc_eventTag = 0;
@@ -4054,6 +4055,12 @@ lpfc_sli4_brdreset(struct lpfc_hba *phba)
  phba->fcf.fcf_flag = 0;
  spin_unlock_irq(&phba->hbalock);
 
+ /* SLI4 INTF 2: if FW dump is being taken skip INIT_PORT */
+ if (phba->hba_flag & HBA_FW_DUMP_OP) {
+  phba->hba_flag &= ~HBA_FW_DUMP_OP;
+  return rc;
+ }
+
  /* Now physically reset the device */
  lpfc_printf_log(phba, KERN_INFO, LOG_INIT,
    "0389 Performing PCI function reset!\n");
-- 
1.7.1